<launch>

  <!-- Launch file for making a turtle move in rviz using turtlebot3_teleop -->

  <!-- Define Robot Arguments -->
  <arg name="robot" default="localhost" doc="Whether launchfile runs the robot in the simulator (localhost) or runs the real turtle (fasturtle.local)"/>
  <arg name="simulate" default="True" doc="Whether launch the simulator and the landmarks node and display everything in rviz"/>

  <!-- Load Turtle Parameters -->
  <rosparam command="load" file="$(find rigid2d)/config/robot_params.yaml" />
  <rosparam command="load" file="$(find nuturtle_description)/config/diff_params.yaml" />
  <rosparam command="load" file="$(find nurtlesim)/config/nurtlesim_params.yaml" />
  <rosparam command="load" file="$(find nuslam)/config/laser_scanner.yaml" />
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(find nuturtle_description)/urdf/my_turtlebot3.urdf.xacro" />

  <!-- launchfile runs the robot in the simulator -->
  <group if="$(arg simulate)">

    <!-- Load turtlebot3_teleop -->
    <include file="$(find turtlebot3_teleop)/launch/turtlebot3_teleop_key.launch" />

    <!-- Load tube_world.cpp Node -->
    <node name="tube_world" pkg="nurtlesim" type="tube_world" output="screen" />

    <!-- Load visualization_node.cpp Node -->
    <node name="visualization" pkg="nuslam" type="visualization" />

    <!-- Load visualization_node.cpp Node -->
    <node name="slam" pkg="nuslam" type="slam" output="screen" />

    <!-- Load landmarks.cpp Node -->
    <node name="landmarks" pkg="nuslam" type="landmarks" output="screen" />

    <!-- Load Robot Model in Rviz -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find nuslam)/config/detect_landmarks.rviz" />
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="rob_st_pub" />
    
  </group>


  <!-- launchfile runs the real turtlebot -->
  <group unless="$(arg simulate)">

  	<!-- Basic Remote Launch -->
		<include file="$(find nuturtle_robot)/launch/basic_remote.launch">
			<arg name="robot" value="$(arg robot)"/>
		</include>

    <!-- Load turtlebot3_teleop -->
    <include file="$(find turtlebot3_teleop)/launch/turtlebot3_teleop_key.launch" />

    <!-- Load fake_turtle_node.cpp Node -->
    <node name="turtle_interface" pkg="nuturtle_robot" type="turtle_interface_node" output="screen" />

    <!-- Load odometer_node.cpp Node -->
    <node name="odometer" pkg="rigid2d" type="odometer_node" output="screen" />

    <!-- Load odometer_node.cpp Node -->
    <node name="tube_world" pkg="nurtlesim" type="tube_world" output="screen" />

    <!-- Load Robot Model in Rviz -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find nuslam)/config/detect_landmarks.rviz" />
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="rob_st_pub" />
    
  </group>

</launch>